api/src/main/java/net/kyori/adventure/animation/util/LoopedListIterator.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * This file is part of adventure, licensed under the MIT License.
+ *
+ * Copyright (c) 2017-2021 KyoriPowered
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package net.kyori.adventure.animation.util;
+
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * A list iterator, that loops after specified index.
+ *
+ * @param <E> type of the iterated elements
+ * @since 4.9.2
+ */
+public final class LoopedListIterator<E> implements Iterator<E> {
```

Comment:
`CircularListIterator`

Reply:
Its not really circular. It may have a linear beginning.

Reply:
No, the iterator is clearly _circular_ in nature. We define _circular_ as a list or iterator whose end points back to the beginning of the list. Suppose the list `L` looks like this:

```
+-----+       +-----+       +-----+
|  1  |  ==>  |  2  |  ==>  |  3  |
+-----+       +-----+       +-----+
```

We define an iterator as a function `I(L) => (N, F(N))` where `N` is the current node to visit and `F(N)` is a function that returns the next node in the list as directed by `I(L)`. Suppose that we use this iterator `I_A(L)` to over over this list multiple times. It is clear that `I_A(L)` wraps another `I(L)`, in fact we can define its behavior like so:

![image](https://user-images.githubusercontent.com/16436212/139584308-41a426ee-cc76-4620-ac49-13fe4c15f43b.png)

The only question is how to define `F_A(F(N))`, and it can be defined as a recursive function:

![image](https://user-images.githubusercontent.com/16436212/139584334-8281a326-4137-4818-ae53-3abba26b793d.png)

We thus conclude this iterator is circular by definition.